### PR TITLE
Fix failing tests from Pydantic v2 migration

### PR DESCRIPTION
**Pull request by hrushikesh-s**
## Summary

* WIP PR to close #548

## Checklist

Work-in-progress pull requests are encouraged, but please put [WIP] in the pull request
title.

Before a pull request can be merged, the following items must be checked:

* [X] Code is in the [standard Python style](https://www.python.org/dev/peps/pep-0008/).
  The easiest way to handle this is to run the following in the **correct sequence** on
  your local machine. Start with running [black](https://black.readthedocs.io/en/stable/index.html) on your new code. This will
  automatically reformat your code to PEP8 conventions and removes most issues. Then run
  [ruff](https://ruff.rs).
* [X] Docstrings have been added in the [Numpy docstring format](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_numpy.html).
  Run [ruff](https://beta.ruff.rs/docs/rules/#pydocstyle-d) on your code.
* [X] Type annotations are **highly** encouraged. Run [mypy](http://mypy-lang.org) to
  type check your code.
* [ ] Tests have been added for any new functionality or bug fixes.
* [ ] All linting and tests pass.

Note that the CI system will run all the above checks. But it will be much more
efficient if you already fix most errors prior to submitting the PR. It is highly
recommended that you use the pre-commit hook provided in the repository. Simply run
`pre-commit install` and a check will be run prior to allowing commits.

### CODE REVIEW

**Comment by hrushikesh-s** [comment]
Hi @janosh ,
Since this PR is to `pydantic` branch of atomate2, the tests are not being run by github upon making the `push` action. Any idea of how to make github run the tests without explicitly changing the [`testing.yaml`](https://github.com/materialsproject/atomate2/blob/pydantic2/.github/workflows/testing.yml) file ?

**Comment by janosh** [comment]
@hrushikesh-s Ah, that's a bit annoying. We should add `workflow_dispatch` as a trigger for manually running tests on select branches. But even that isn't ideal, so maybe we should get rid of the `branches: [main]` filter for PRs.

**Comment by janosh** [comment]
Try pulling 12eacf0a into your branch.

File: src/atomate2/cp2k/schemas/calculation.py [hunk 1]
```diff
@@ -82,16 +82,16 @@ class CalculationInput(BaseModel):
         description="CP2K global parameters used in the last calc of this task.",
     )
 
-    @validator("atomic_kind_info")
-    def remove_unnecessary(self, atomic_kind_info):
+    @field_validator("atomic_kind_info")
+    def remove_unnecessary(cls, atomic_kind_info):
```
Comment (janosh) [inline src/atomate2/cp2k/schemas/calculation.py:86]: `self` should stay as is.

**Comment by hrushikesh-s** [comment]
@janosh, I'm getting the following assertion error for `test_m3gnet_static_maker` & `test_m3gnet_relax_maker`

```
FAILED tests/forcefields/test_jobs.py::test_m3gnet_static_maker - assert -10.711267471313477 == -10.8454 ± 1.1e-03
FAILED tests/forcefields/test_jobs.py::test_m3gnet_relax_maker - assert -10.710837364196777 == -10.8441 ± 1.1e-03
```

**Comment by janosh** [comment]
Weird that the values are not consistent. Feel free to loosen for now:

```pytest
assert output1.output.energy == approx(-10.8, abs=0.2)
```

**Comment by hrushikesh-s** [comment]
@janosh , I'm getting the following errors while running [`test_nonrad_maker`](https://github.com/materialsproject/atomate2/blob/pydantic2/tests/vasp/flows/test_defect.py#L63-L134) & [`test_formation_energy_maker`](https://github.com/materialsproject/atomate2/blob/pydantic2/tests/vasp/flows/test_defect.py#L137-L191) function from [`tests/vasp/flows/test_defect.py`](https://github.com/materialsproject/atomate2/blob/pydantic2/tests/vasp/flows/test_defect.py)

```
File "/.../atomate2/src/atomate2/common/jobs/defect.py", line 401, in check_charge_state
    raise ValueError(
ValueError: The charge of the output structure is 0, but expected charge state from the Defect object is -2.

File "/.../atomate2/src/atomate2/common/jobs/defect.py", line 401, in check_charge_state
    raise ValueError(
ValueError: The charge of the output structure is 0, but expected charge state from the Defect object is -1.
```

**Comment by janosh** [comment]
@hrushikesh-s Not sure what changed. Maybe @jmmshn has troubleshooting suggestions?

**Comment by hrushikesh-s** [comment]
@janosh, there's one last error that's coming up in `test_formation_energy_maker` function in `test_defect.py`:

```
File "/.../atomate2/src/atomate2/common/jobs/defect.py", line 416, in get_defect_entry
    energy=bulk_c_entry.energy,
AttributeError: 'NoneType' object has no attribute 'energy'
```

**Comment by jmmshn** [comment]
Hi @janosh, It looks like there might have been some upstream changes with parsing the net charge of structures from a VASP directory.  The code here was just to do a sanity check of the structure charge before things enter the DB.
I'll take a look at this by Wednesday.

**Comment by jmmshn** [comment]
So I just checked the `pymatgen` code for parsing the structure charge and I don't think anything has changed.  It might be helpful to just run `Vasprun(<file_path>, parse_potcar_file=True)` on the directories that are causing problems.

**Comment by hrushikesh-s** [comment]
So, do we need to make any changes to [deploy.yaml](https://github.com/materialsproject/atomate2/blob/main/.github/workflows/deploy.yml)?

**Comment by janosh** [comment]
@jmmshn Thanks for digging into this so quickly! Re unusable POTCARs in CI, maybe something to simulate with `from unittest.mock import patch`? Either way, that's for a future PR.

@hrushikesh-s Thanks a lot for your help here! This was really helpful! I'll merge this PR and take a look at [the last error](https://github.com/materialsproject/atomate2/pull/558#issuecomment-1754077291) over in #558.

**Comment by jmmshn** [comment]
I don't think so.  If CI is actually breaking on this then we will just have to fix the code.  I'll mark some time on Wednesday to make sure this all works.

**Comment by jmmshn** [comment]
@janosh, yeah I think I just ignored it because this failed job was passing CI and it's kind of nice to see the fact that the check is working.  Either way I should probably:
- make one of the defect charges parse properly by mocking
- make the other one fail but process that exception better in the testing suite so it doesn't cause this exact confusion.

**Comment by janosh** [comment]
> * make one of the defect charges parse properly by mocking
> * make the other one fail but process that exception better in the testing suite so it doesn't cause this exact confusion.

That would be perfect!

**Pull request merged**